File: python/pysearch/filetypes.py
# -*- coding: utf-8 -*-
################################################################################
#
# fileutil.py
#
# class FileUtil: provides utility functions for getting file extension and
#                 determining file type
#
################################################################################
import os
import xml.dom.minidom as minidom

from common import get_text
from config import FILETYPESPATH
from fileutil import FileUtil

class FileType(object):
    """FileType enum"""
    Archive = 1
    Binary  = 2
    Text    = 3
    Unknown = 4

class FileTypes(object):
    """a class to provide file type information"""

    def __init__(self, **kargs):
        self.filetypes = {}
        self.__dict__.update(kargs)
        self.__populate_filetypes()

    def get_filetype(self, filename):
        if self.is_text_file(filename):
            return FileType.Text
        elif self.is_binary_file(filename):
            return FileType.Binary
        elif self.is_archive_file(filename):
            return FileType.Archive
        else:
            return FileType.Unknown

    def is_archive_file(self, f):
        """Return true if file is of a (known) archive file type"""
        return FileUtil.get_extension(f) in self.filetypes['archive']

    def is_binary_file(self, f):
        """Return true if file is of a (known) searchable binary file type"""
        return FileUtil.get_extension(f) in self.filetypes['binary']

    def is_searchable_file(self, f):
        """Return true if file is of a (known) searchable type"""
        return FileUtil.get_extension(f) in self.filetypes['searchable']

    def is_text_file(self, f):
        """Return true if file is of a (known) text file type"""
        return FileUtil.get_extension(f) in self.filetypes['text']

    def __populate_filetypes(self):
        filetypedom = minidom.parse(FILETYPESPATH)
        filetypenodes = filetypedom.getElementsByTagName('filetype')
        for filetypenode in filetypenodes:
            name = filetypenode.getAttribute('name')
            extnode = filetypenode.getElementsByTagName('extensions')[0]
            exts = set(get_text(extnode.childNodes).split())
            self.filetypes[name] = exts
        self.filetypes['text'] = \
            self.filetypes['text'].union(self.filetypes['code'],
                                         self.filetypes['xml'])
        self.filetypes['searchable'] = \
            self.filetypes['binary'].union(self.filetypes['archive'],
                                           self.filetypes['text'])

File: node/nodesearch/src/filetypes.js
/*
 * filetypes.js
 *
 * identifies file types (archive, binary, text, unknown)
 */

var fs = require('fs');

var common = require('./common.js');
var config = require('./config.js');
var FileType = require('./filetype.js').FileType;
var FileUtil = require('./fileutil.js').FileUtil;

function FileTypes() {
    var self = this;

    var fileTypeMap = (function () {
        //TODO: move to config file
        var fileTypeMap = {};

        var DomJS = require("dom-js").DomJS;
        var domjs = new DomJS();

        var xml = fs.readFileSync(FileUtil.expandPath(config.FILETYPESPATH)).toString();
        domjs.parse(xml, function(err, dom) {
            if (err) {
                throw err;
            }
            for (var i in dom.children) {
                var child = dom.children[i];
                if (child.name && child.name === 'filetype') {
                    var name = child.attributes.name;
                    for (var j in child.children) {
                        var filetypeChild = child.children[j];
                        if (filetypeChild.name && filetypeChild.name === 'extensions') {
                            var extensions = filetypeChild.children[0].text;
                            fileTypeMap[name] = common.setFromArray(extensions.split(/\s+/));
                        }
                    }
                }
            }
            fileTypeMap.text = [].concat(fileTypeMap.text, fileTypeMap.code, fileTypeMap.xml);
            fileTypeMap.searchable = [].concat(fileTypeMap.text, fileTypeMap.binary,
                fileTypeMap.archive);
        });
        return fileTypeMap;
    })();

    this.getFileType = function (filename) {
        if (self.isTextFile(filename))
            return FileType.TEXT;
        if (self.isBinaryFile(filename))
            return FileType.BINARY;
        if (self.isArchiveFile(filename))
            return FileType.ARCHIVE;
        return FileType.UNKNOWN;
    };

    this.isArchiveFile = function (filename) {
        var ext = FileUtil.getExtension(filename);
        return fileTypeMap.archive.indexOf(ext) > -1;
    };

    this.isBinaryFile = function (filename) {
        var ext = FileUtil.getExtension(filename);
        return fileTypeMap.binary.indexOf(ext) > -1;
    };

    this.isSearchableFile = function (filename) {
        var ext = FileUtil.getExtension(filename);
        return fileTypeMap.searchable.indexOf(ext) > -1;
    };

    this.isTextFile = function (filename) {
        var ext = FileUtil.getExtension(filename);
        return fileTypeMap.text.indexOf(ext) > -1;
    };

    this.isUnknownFile = function (filename) {
        var ext = FileUtil.getExtension(filename);
        return fileTypeMap.searchable.indexOf(ext) === -1;
    };
}

exports.FileTypes = FileTypes;